src/ott/solvers/nn/neuraldual.py

```diff
@@ -96,66 +153,195 @@ def __init__(
+      self.train_step_g = self.get_step_fn(train=True, to_optimize="g")
+      self.valid_step_g = self.get_step_fn(train=False, to_optimize="g")
+      self.train_fn = self.train_neuraldual_alternating
 
   def __call__(
       self,
       trainloader_source: Iterable[jnp.ndarray],
       trainloader_target: Iterable[jnp.ndarray],
       validloader_source: Iterable[jnp.ndarray],
       validloader_target: Iterable[jnp.ndarray],
+      callback: Optional[Callback_t] = None,
   ) -> Union[potentials.DualPotentials, Tuple[potentials.DualPotentials,
                                               Train_t]]:
-    logs = self.train_neuraldual(
+    logs = self.train_fn(
         trainloader_source,
         trainloader_target,
         validloader_source,
         validloader_target,
+        callback=callback,
     )
     res = self.to_dual_potentials()
 
     return (res, logs) if self.logging else res
 
-  def train_neuraldual(
+  def train_neuraldual_parallel(
+      self,
+      trainloader_source: Iterable[jnp.ndarray],
+      trainloader_target: Iterable[jnp.ndarray],
+      validloader_source: Iterable[jnp.ndarray],
+      validloader_target: Iterable[jnp.ndarray],
+      callback: Optional[Callback_t] = None,
+  ) -> Train_t:
+    """Implementation of the training and validation with parallel updates."""  # noqa: D401
+    try:
+      from tqdm.auto import tqdm
+    except ImportError:
+      tqdm = lambda _: _
+    # define dict to contain source and target batch
+    train_batch = {}
+    valid_batch = {}
+
+    # set logging dictionaries
+    train_logs = {
+        "train_loss_f": [],
+        "train_loss_g": [],
+        "train_w_dist": [],
+        "directions": []
+    }
+    valid_logs = {"valid_loss_f": [], "valid_loss_g": [], "valid_w_dist": []}
+
+    for step in tqdm(range(self.num_train_iters)):
+      if not self.back_and_forth or step % 2 == 0:
+        # Update the forward direction
+        train_batch["source"] = jnp.asarray(next(trainloader_source))
+        train_batch["target"] = jnp.asarray(next(trainloader_target))
+        self.state_f, self.state_g, loss, loss_f, loss_g, w_dist = self.train_step_parallel(
+            self.state_f,
+            self.state_g,
+            train_batch,
+        )
+
+        if self.logging and step % self.log_freq == 0:
+          train_logs["train_loss_f"].append(float(loss_f))
+          train_logs["train_loss_g"].append(float(loss_g))
+          train_logs["train_w_dist"].append(float(w_dist))
+          train_logs["directions"].append('forward')
+      else:
+        # Update the backward direction
+        train_batch["target"] = jnp.asarray(next(trainloader_source))
+        train_batch["source"] = jnp.asarray(next(trainloader_target))
+        self.state_g, self.state_f, loss, loss_f, loss_g, w_dist = self.train_step_parallel(
+            self.state_g,
+            self.state_f,
+            train_batch,
+        )
+
+        if self.logging and step % self.log_freq == 0:
```

Comment:
There's too much code duplication when updating the logs, would move it outside the `if` as
```python
train_logs["train_loss_f"].append(float(loss_f))
train_logs["train_loss_g"].append(float(loss_g))
train_logs["train_w_dist"].append(float(w_dist))
train_logs["directions"].append('forward' if update_forward else 'backward')
```